cranelift/codegen/src/egraph/stores.rs

```diff
@@ -0,0 +1,270 @@
+//! hashconsing. We design the Eq/Hash (node identity relation
+//! definition) of the `Node` struct so that all loads with (i) the
+//! same "last store", and (ii) the same address expression, and (iii)
+//! the same opcode-and-offset, will deduplicate (the first will be
+//! computed, and the later ones will use the same value). Furthermore
+//! we have a rewrite rule (in opts/store_to_load.isle) that rewrites
+//! a load into the stored value of the last store *if* the last store
+//! has the same address expression and constant offset.
+//!
+//! This gives us two optimizations, "redundant load elimination" and
+//! "store-to-load forwarding".
+//!
+//! In theory we could also do *dead-store elimination*, where if a
+//! store overwrites a value earlier written by another store, *and*
+//! if no other load/store to the abstract state category occurred,
+//! *and* no other trapping instruction occurred (at which point we
+//! need an up-to-date memory state because post-trap-termination
+//! memory state can be observed), *and* we can prove the original
+//! store could not have trapped, then we can eliminate the original
+//! store. Because this is so complex, and the conditions for doing it
+//! correctly when post-trap state must be correct likely reduce the
+//! potential benefit, we don't yet do this.
+
+use crate::flowgraph::ControlFlowGraph;
+use crate::fx::{FxHashMap, FxHashSet};
+use crate::inst_predicates::has_memory_fence_semantics;
+use crate::ir::{Block, Function, Inst, Opcode};
+use cranelift_entity::EntityRef;
+use cranelift_entity::SecondaryMap;
+use smallvec::{smallvec, SmallVec};
+
+/// For a given program point, the vector of last-store instruction
+/// indices for each disjoint category of abstract state.
+#[derive(Clone, Copy, Debug, Default, PartialEq, Eq)]
+struct LastStores {
+    heap: MemoryState,
+    table: MemoryState,
+    vmctx: MemoryState,
+    other: MemoryState,
+}
+
+/// State of memory seen by a load.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord, Hash)]
+pub enum MemoryState {
+    Entry,
+    Store(Inst),
+    BeforeInst(Inst),
+    AfterInst(Inst),
+    Top,
+}
+
+impl std::default::Default for MemoryState {
+    fn default() -> Self {
+        MemoryState::Top
+    }
+}
+
+impl LastStores {
+    fn entry() -> Self {
+        Self {
+            heap: MemoryState::Entry,
+            table: MemoryState::Entry,
+            vmctx: MemoryState::Entry,
+            other: MemoryState::Entry,
+        }
+    }
+
+    fn update(&mut self, func: &Function, inst: Inst) {
+        let opcode = func.dfg[inst].opcode();
+        if has_memory_fence_semantics(opcode) {
+            self.heap = MemoryState::AfterInst(inst);
+            self.table = MemoryState::AfterInst(inst);
+            self.vmctx = MemoryState::AfterInst(inst);
+            self.other = MemoryState::AfterInst(inst);
+        } else if opcode.can_store() {
+            if let Some(memflags) = func.dfg[inst].memflags() {
```

Comment:
> [handle stack stores specially]

Yep, we definitely could do this. None of the benchmarks I had been developing with use them (actually, `cranelift-wasm` doesn't use them at all) so I didn't have an easy way to test or evaluate the effect of this; it can definitely be done as future work.

> [treat some instructions only as "other"]

Unfortunately I don't think this is correct in all circumstances: an instruction about which we don't know anything has to clobber all segments of the state, not just "other" (i.e., it could alias with any of the categories).

It's possible that reasoning about the specific instruction formats that exist could lead us to something more precise here, but I really don't want to go down that road, because it bakes in knowledge of the opcode set as static and unchanging, and thus sets up an action-at-a-distance timebomb for when we later add some other instruction kind that involves memory. I'd rather have fallback logic here that basically says "if we don't know, clobber everything" -- it's the safe and obviously-correct (for some value of "obvious") approach :-)

Reply:
Somehow I assumed cranelift-wasm used stack slots, but I guess there really is no reason for it to. Huh!

Well, I don't think it matters for this PR, but perhaps `has_memory_fence_semantics` should still be updated to include `Debugtrap`? It does seem like an instruction that shouldn't have memory accesses reordered across it.

Reply:
Ah, yes, I agree -- updated (and actually added `op if op.can_trap()` as an additional match arm).